src/Statement/Grouper/FilteringStatementGrouper.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+namespace Wikibase\DataModel\Services\Statement\Grouper;
+
+use InvalidArgumentException;
+use Wikibase\DataModel\Statement\Statement;
+use Wikibase\DataModel\Statement\StatementFilter;
+use Wikibase\DataModel\Statement\StatementList;
+
+/**
+ * @since 3.2
+ *
+ * @licence GNU GPL v2+
+ * @author Thiemo Mättig
+ */
+class FilteringStatementGrouper implements StatementGrouper {
```

Comment:
Funny, I'm currently considering removing the grouping from the JS code and the JSON in `wbEntity`. To me, it's a pain to work with the grouped statement list, since we are now going to look at them by GUID alone.

As for statement groupers: The fact that this class is non-trivial seems convincing enough to me. Note that using non-exclusive filters for grouping is not that obvious or trivial. We might just as well have groupers that don't work with filters.